File: backend/app/schemas/quality.py
from __future__ import annotations
from datetime import date, datetime
from decimal import Decimal
from typing import Optional, List
import uuid

from pydantic import BaseModel, ConfigDict, computed_field

from app.models.quality import QCType, QCStatus, QCDecision, ParameterType


# ── QC Parameter ──────────────────────────────────────────────────────────────

class QCParameterCreate(BaseModel):
    name: str
    description: Optional[str] = None
    parameter_type: ParameterType = ParameterType.NUMERIC
    unit: Optional[str] = None
    min_value: Optional[Decimal] = None
    max_value: Optional[Decimal] = None
    expected_value: Optional[str] = None
    is_critical: bool = False
    applicable_types: str = "ALL"


class QCParameterUpdate(BaseModel):
    description: Optional[str] = None
    unit: Optional[str] = None
    min_value: Optional[Decimal] = None
    max_value: Optional[Decimal] = None
    expected_value: Optional[str] = None
    is_critical: Optional[bool] = None
    applicable_types: Optional[str] = None
    is_active: Optional[bool] = None


class QCParameterRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: uuid.UUID
    name: str
    description: Optional[str]
    parameter_type: ParameterType
    unit: Optional[str]
    min_value: Optional[Decimal]
    max_value: Optional[Decimal]
    expected_value: Optional[str]
    is_critical: bool
    applicable_types: str
    is_active: bool


# ── Test Results ──────────────────────────────────────────────────────────────

class QCTestResultCreate(BaseModel):
    parameter_id: Optional[uuid.UUID] = None
    parameter_name: str
    parameter_type: ParameterType
    unit: Optional[str] = None
    numeric_value: Optional[Decimal] = None
    text_value: Optional[str] = None
    boolean_value: Optional[bool] = None
    min_spec: Optional[Decimal] = None
    max_spec: Optional[Decimal] = None
    expected_text: Optional[str] = None
    is_passed: Optional[bool] = None
    is_critical: bool = False
    notes: Optional[str] = None


class QCTestResultRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: uuid.UUID
    inspection_id: uuid.UUID
    parameter_id: Optional[uuid.UUID]
    parameter_name: str
    parameter_type: ParameterType
    unit: Optional[str]
    numeric_value: Optional[Decimal]
    text_value: Optional[str]
    boolean_value: Optional[bool]
    min_spec: Optional[Decimal]
    max_spec: Optional[Decimal]
    expected_text: Optional[str]
    is_passed: Optional[bool]
    is_critical: bool
    tested_by_id: Optional[uuid.UUID]
    tested_by_name: Optional[str] = None
    notes: Optional[str]
    created_at: datetime

    @computed_field
    @property
    def display_value(self) -> str:
        if self.numeric_value is not None:
            return f"{self.numeric_value} {self.unit or ''}".strip()
        if self.text_value is not None:
            return self.text_value
        if self.boolean_value is not None:
            return "Yes" if self.boolean_value else "No"
        return "—"

    @computed_field
    @property
    def spec_range(self) -> Optional[str]:
        if self.min_spec is not None and self.max_spec is not None:
            return f"{self.min_spec} – {self.max_spec} {self.unit or ''}".strip()
        if self.min_spec is not None:
            return f"≥ {self.min_spec} {self.unit or ''}".strip()
        if self.max_spec is not None:
            return f"≤ {self.max_spec} {self.unit or ''}".strip()
        if self.expected_text:
            return self.expected_text
        return None


# ── QC Inspection ─────────────────────────────────────────────────────────────

class QCInspectionCreate(BaseModel):
    inspection_no: str
    qc_type: QCType
    grn_id: Optional[uuid.UUID] = None
    grn_line_id: Optional[uuid.UUID] = None
    production_order_id: Optional[uuid.UUID] = None
    lot_id: Optional[uuid.UUID] = None
    supplier_id: Optional[uuid.UUID] = None
    material_id: Optional[uuid.UUID] = None
    product_id: Optional[uuid.UUID] = None
    warehouse_id: Optional[uuid.UUID] = None
    lot_number: Optional[str] = None
    batch_no: Optional[str] = None
    sample_size: Optional[Decimal] = None
    sample_unit: Optional[str] = None
    inspection_date: date
    notes: Optional[str] = None


class QCInspectionUpdate(BaseModel):
    inspection_date: Optional[date] = None
    sample_size: Optional[Decimal] = None
    sample_unit: Optional[str] = None
    notes: Optional[str] = None


class QCDecisionRequest(BaseModel):
    decision: QCDecision
    decision_notes: Optional[str] = None
    rework_candidate: bool = False
    apply_quarantine: bool = False


class QCInspectionRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: uuid.UUID
    inspection_no: str
    qc_type: QCType
    status: QCStatus
    grn_id: Optional[uuid.UUID]
    production_order_id: Optional[uuid.UUID]
    lot_id: Optional[uuid.UUID]
    supplier_id: Optional[uuid.UUID]
    supplier_name: Optional[str] = None
    material_id: Optional[uuid.UUID]
    material_name: Optional[str] = None
    material_code: Optional[str] = None
    product_id: Optional[uuid.UUID]
    product_name: Optional[str] = None
    product_sku: Optional[str] = None
    warehouse_id: Optional[uuid.UUID]
    lot_number: Optional[str]
    batch_no: Optional[str]
    sample_size: Optional[Decimal]
    sample_unit: Optional[str]
    inspection_date: date
    inspector_id: Optional[uuid.UUID]
    inspector_name: Optional[str] = None
    notes: Optional[str]
    decision: Optional[QCDecision]
    decided_by_id: Optional[uuid.UUID]
    decided_at: Optional[datetime]
    decision_notes: Optional[str]
    rework_candidate: bool
    quarantine_applied: bool
    created_at: datetime
    test_count: int = 0
    pass_count: int = 0
    fail_count: int = 0
    critical_fail: bool = False


class QCInspectionDetailRead(QCInspectionRead):
    test_results: List[QCTestResultRead] = []


# ── COA (Certificate of Analysis) ─────────────────────────────────────────────

class COATestLine(BaseModel):
    parameter_name: str
    unit: Optional[str]
    spec_range: Optional[str]
    result_value: str
    is_passed: Optional[bool]
    is_critical: bool


class COAReport(BaseModel):
    inspection_no: str
    qc_type: str
    inspection_date: date
    lot_number: Optional[str]
    batch_no: Optional[str]
    product_name: Optional[str]
    product_sku: Optional[str]
    material_name: Optional[str]
    supplier_name: Optional[str]
    inspector_name: Optional[str]
    decision: Optional[str]
    decision_notes: Optional[str]
    decided_at: Optional[datetime]
    overall_result: str   # PASS / FAIL / CONDITIONAL / PENDING
    test_lines: List[COATestLine]


# ── Stats / Reports ────────────────────────────────────────────────────────────

class QCStatsRow(BaseModel):
    period: str
    total_inspections: int
    passed: int
    failed: int
    conditional: int
    pass_rate: float


class RejectionSummaryRow(BaseModel):
    material_name: Optional[str]
    product_name: Optional[str]
    supplier_name: Optional[str]
    total_inspections: int
    total_rejected: int
    rejection_rate: float
    last_rejection_date: Optional[date]

File: frontend/src/lib/quality.ts
import { apiClient } from "@/lib/api";

export type QCType = "INCOMING" | "IN_PROCESS" | "FINISHED_GOODS";
export type QCStatus = "PENDING" | "IN_PROGRESS" | "PASSED" | "FAILED" | "CONDITIONAL_RELEASE" | "CANCELLED";
export type QCDecision = "ACCEPT" | "REJECT" | "CONDITIONAL_RELEASE" | "REWORK";
export type ParameterType = "NUMERIC" | "TEXT" | "BOOLEAN" | "PASS_FAIL";

export interface QCParameter {
  id: string;
  name: string;
  description?: string;
  parameter_type: ParameterType;
  unit?: string;
  min_value?: number;
  max_value?: number;
  expected_value?: string;
  is_critical: boolean;
  applicable_types: string;
  is_active: boolean;
}

export interface QCTestResult {
  id: string;
  inspection_id: string;
  parameter_id?: string;
  parameter_name: string;
  parameter_type: ParameterType;
  unit?: string;
  numeric_value?: number;
  text_value?: string;
  boolean_value?: boolean;
  min_spec?: number;
  max_spec?: number;
  expected_text?: string;
  is_passed?: boolean;
  is_critical: boolean;
  tested_by_id?: string;
  tested_by_name?: string;
  notes?: string;
  created_at: string;
  display_value: string;
  spec_range?: string;
}

export interface QCInspection {
  id: string;
  inspection_no: string;
  qc_type: QCType;
  status: QCStatus;
  grn_id?: string;
  production_order_id?: string;
  lot_id?: string;
  supplier_id?: string;
  supplier_name?: string;
  material_id?: string;
  material_name?: string;
  material_code?: string;
  product_id?: string;
  product_name?: string;
  product_sku?: string;
  warehouse_id?: string;
  lot_number?: string;
  batch_no?: string;
  sample_size?: number;
  sample_unit?: string;
  inspection_date: string;
  inspector_id?: string;
  inspector_name?: string;
  notes?: string;
  decision?: QCDecision;
  decided_by_id?: string;
  decided_at?: string;
  decision_notes?: string;
  rework_candidate: boolean;
  quarantine_applied: boolean;
  created_at: string;
  test_count: number;
  pass_count: number;
  fail_count: number;
  critical_fail: boolean;
}

export interface QCInspectionDetail extends QCInspection {
  test_results: QCTestResult[];
}

export interface COATestLine {
  parameter_name: string;
  unit?: string;
  spec_range?: string;
  result_value: string;
  is_passed?: boolean;
  is_critical: boolean;
}

export interface COAReport {
  inspection_no: string;
  qc_type: string;
  inspection_date: string;
  lot_number?: string;
  batch_no?: string;
  product_name?: string;
  product_sku?: string;
  material_name?: string;
  supplier_name?: string;
  inspector_name?: string;
  decision?: string;
  decision_notes?: string;
  decided_at?: string;
  overall_result: string;
  test_lines: COATestLine[];
}

export interface RejectionSummaryRow {
  material_name?: string;
  product_name?: string;
  supplier_name?: string;
  total_inspections: number;
  total_rejected: number;
  rejection_rate: number;
  last_rejection_date?: string;
}

export const qualityApi = {
  // Parameters
  async listParameters(params?: { qc_type?: QCType; active_only?: boolean }): Promise<QCParameter[]> {
    const res = await apiClient.get<QCParameter[]>("/api/v1/quality/parameters/", { params });
    return res.data;
  },
  async createParameter(data: object): Promise<QCParameter> {
    const res = await apiClient.post<QCParameter>("/api/v1/quality/parameters/", data);
    return res.data;
  },
  async updateParameter(id: string, data: object): Promise<QCParameter> {
    const res = await apiClient.patch<QCParameter>(`/api/v1/quality/parameters/${id}`, data);
    return res.data;
  },
  async deleteParameter(id: string): Promise<void> {
    await apiClient.delete(`/api/v1/quality/parameters/${id}`).then((r) => r.data);
  },

  // Inspections
  async listInspections(params?: {
    qc_type?: QCType;
    status?: QCStatus;
    supplier_id?: string;
    material_id?: string;
    product_id?: string;
  }): Promise<QCInspection[]> {
    const res = await apiClient.get<QCInspection[]>("/api/v1/quality/inspections/", { params });
    return res.data;
  },
  async getInspection(id: string): Promise<QCInspectionDetail> {
    const res = await apiClient.get<QCInspectionDetail>(`/api/v1/quality/inspections/${id}`);
    return res.data;
  },
  async createInspection(data: object): Promise<QCInspectionDetail> {
    const res = await apiClient.post<QCInspectionDetail>("/api/v1/quality/inspections/", data);
    return res.data;
  },
  async updateInspection(id: string, data: object): Promise<QCInspectionDetail> {
    const res = await apiClient.patch<QCInspectionDetail>(`/api/v1/quality/inspections/${id}`, data);
    return res.data;
  },
  async addTestResult(inspectionId: string, data: object): Promise<QCInspectionDetail> {
    const res = await apiClient.post<QCInspectionDetail>(`/api/v1/quality/inspections/${inspectionId}/results`, data);
    return res.data;
  },
  async deleteTestResult(inspectionId: string, resultId: string): Promise<void> {
    await apiClient.delete(`/api/v1/quality/inspections/${inspectionId}/results/${resultId}`).then((r) => r.data);
  },
  async decide(inspectionId: string, data: object): Promise<QCInspectionDetail> {
    const res = await apiClient.post<QCInspectionDetail>(`/api/v1/quality/inspections/${inspectionId}/decide`, data);
    return res.data;
  },
  async releaseQuarantine(inspectionId: string, notes?: string): Promise<QCInspectionDetail> {
    const res = await apiClient.post<QCInspectionDetail>(
      `/api/v1/quality/inspections/${inspectionId}/release-quarantine`,
      null,
      { params: notes ? { notes } : {} }
    );
    return res.data;
  },
  async getCOA(inspectionId: string): Promise<COAReport> {
    const res = await apiClient.get<COAReport>(`/api/v1/quality/inspections/${inspectionId}/coa`);
    return res.data;
  },
  async createForGRN(params: {
    grn_id: string; grn_line_id: string; inspection_no: string;
    material_id?: string; supplier_id?: string; warehouse_id?: string; lot_number?: string;
  }): Promise<QCInspectionDetail> {
    const res = await apiClient.post<QCInspectionDetail>("/api/v1/quality/create-for-grn", null, { params });
    return res.data;
  },
  async createForOrder(params: {
    production_order_id: string; inspection_no: string;
    product_id?: string; warehouse_id?: string; lot_number?: string; batch_no?: string;
    qc_type?: QCType;
  }): Promise<QCInspectionDetail> {
    const res = await apiClient.post<QCInspectionDetail>("/api/v1/quality/create-for-order", null, { params });
    return res.data;
  },

  // Reports
  async rejectionSummary(): Promise<RejectionSummaryRow[]> {
    const res = await apiClient.get<RejectionSummaryRow[]>("/api/v1/quality/reports/rejections");
    return res.data;
  },
};
